build/lib/release.sh

```diff
@@ -393,22 +393,21 @@ EOF
             docker_build_opts+=("--pull")
         fi
         "${DOCKER[@]}" build "${docker_build_opts[@]}" -q -t "${docker_image_tag}" "${docker_build_path}" >/dev/null
-        "${DOCKER[@]}" save "${docker_image_tag}" > "${binary_dir}/${binary_name}.tar"
+        # If we are building an official/alpha/beta release we want to keep
+        # docker images and tag them appropriately.
+        local release_docker_image_tag=""
+        if [[ -n "${KUBE_DOCKER_IMAGE_TAG-}" && -n "${KUBE_DOCKER_REGISTRY-}" && $docker_registry != $KUBE_DOCKER_REGISTRY ]]; then
+          release_docker_image_tag="${KUBE_DOCKER_REGISTRY}/${binary_name}-${arch}:${KUBE_DOCKER_IMAGE_TAG}"
+          kube::log::status "Tagging docker image ${docker_image_tag} as ${release_docker_image_tag}"
+          "${DOCKER[@]}" rmi "${release_docker_image_tag}" 2>/dev/null || true
+          "${DOCKER[@]}" tag "${docker_image_tag}" "${release_docker_image_tag}" 2>/dev/null
+        fi
+        "${DOCKER[@]}" save "${docker_image_tag}" "${release_docker_image_tag}" -o "${binary_dir}/${binary_name}.tar"
         echo "${docker_tag}" > "${binary_dir}/${binary_name}.docker_tag"
         rm -rf "${docker_build_path}"
         ln "${binary_dir}/${binary_name}.tar" "${images_dir}/"
 
-        # If we are building an official/alpha/beta release we want to keep
-        # docker images and tag them appropriately.
-        if [[ -n "${KUBE_DOCKER_IMAGE_TAG-}" && -n "${KUBE_DOCKER_REGISTRY-}" ]]; then
-          local release_docker_image_tag="${KUBE_DOCKER_REGISTRY}/${binary_name}-${arch}:${KUBE_DOCKER_IMAGE_TAG}"
-          # Only rmi and tag if name is different
-          if [[ $docker_image_tag != $release_docker_image_tag ]]; then
-            kube::log::status "Tagging docker image ${docker_image_tag} as ${release_docker_image_tag}"
-            "${DOCKER[@]}" rmi "${release_docker_image_tag}" 2>/dev/null || true
-            "${DOCKER[@]}" tag "${docker_image_tag}" "${release_docker_image_tag}" 2>/dev/null
-          fi
-        else
+        if [[ -z "${KUBE_DOCKER_IMAGE_TAG-}" || -z "${KUBE_DOCKER_REGISTRY-}" ]]; then
```

Comment:
This line looks duplicated?

Reply:
correct me if I'm wrong here:
We tag the image with k8s.gcr.io + the KUBE_DOCKER_REGISTRY (if defined), but we want to keep images only when we do releases.

I think we can remove the tag docker_image_tag all the time (or maybe when release_docker_image_tag is defined), since the image was already exported (docker save) with that tag associated and the release scripts are using the archives (docker load) to push the images.
https://github.com/kubernetes/release/blob/master/lib/releaselib.sh#L911

What do you think?

Reply:
Actually, I was wrong when I read this.. I thought you were using -n, but you're using -z to flip the logic (this would have been the else statement). So this appears to maintain current functionality. I'm okay with just leaving this as-is.